polygeist/lib/polygeist/Passes/ParallelLICM.cpp

```diff
@@ -371,27 +544,189 @@ void moveParallelLoopInvariantCode(AffineParallelOp loop) {
+
+    ub = ub.replaceDimsAndSymbols(dims, symbols);
+
+    for (AffineExpr lb : lbMap.getResults()) {
+      // Bound is whether this expr >= 0, which since we want ub > lb,
+      // we rewrite as follows.
+      exprs.push_back(ub - lb - loop.getStep());
+      eqflags.push_back(false);
+    }
+  }
+
+  SmallVector<Value> values;
+  OperandRange lb_ops = loop.getLowerBoundOperands(),
+               ub_ops = loop.getUpperBoundOperands();
+
+  std::copy(lb_ops.begin(),
+            lb_ops.begin() + loop.getLowerBoundMap().getNumDims(),
+            std::back_inserter(values));
+  std::copy(ub_ops.begin(),
+            ub_ops.begin() + loop.getUpperBoundMap().getNumDims(),
+            std::back_inserter(values));
+  std::copy(lb_ops.begin() + loop.getLowerBoundMap().getNumDims(), lb_ops.end(),
+            std::back_inserter(values));
+  std::copy(ub_ops.begin() + loop.getUpperBoundMap().getNumDims(), ub_ops.end(),
+            std::back_inserter(values));
+
+  auto iset = IntegerSet::get(
+      /*dim*/ lbMap.getNumDims() + ubMap.getNumDims(),
+      /*symbols*/ lbMap.getNumSymbols() + ubMap.getNumSymbols(), exprs,
+      eqflags);
+
+  bool yieldsResults = !loop.getResults().empty();
+  TypeRange types(loop.getResults());
+  auto ifOp = b.create<AffineIfOp>(loop.getLoc(), types, iset, values,
+                                   yieldsResults ? true : false);
+
+  if (yieldsResults) {
+    ifOp.getThenBodyBuilder().create<AffineYieldOp>(loop.getLoc(),
+                                                    loop.getResults());
+    ifOp.getElseBodyBuilder().create<AffineYieldOp>(loop.getLoc(),
+                                                    loop.getIterOperands());
+  }
+
+  loop->moveBefore(ifOp.getThenBlock()->getTerminator());
+
+  // Replace uses of the loop return value(s) with the value(s) yielded by the
+  // if operation.
+  for (auto it : llvm::zip(loop.getResults(), ifOp.getResults()))
+    std::get<0>(it).replaceUsesWithIf(std::get<1>(it), [&](OpOperand &op) {
+      Block *useBlock = op.getOwner()->getBlock();
+      return useBlock != ifOp.getThenBlock();
+    });
+}
+
+static size_t moveLoopInvariantCode(LoopLikeOpInterface loop,
+                                    AliasAnalysis &aliasAnalysis) {
+  Operation *loopOp = loop;
+  if (!isa<scf::ParallelOp, AffineParallelOp, AffineForOp>(loopOp))
+    return 0;
+
+  SmallVector<Operation *, 8> opsToMove;
+  collectHoistableOperations(loop, aliasAnalysis, opsToMove);
+  if (opsToMove.empty())
+    return 0;
+
+  bool guardedLoop =
+      TypeSwitch<Operation *, bool>(loopOp)
+          .Case<scf::ParallelOp, AffineParallelOp, AffineForOp>([&](auto loop) {
+            createLoopGuard(loop);
+            return true;
+          })
+          .Default([](auto) { return false; });
+
+  size_t numOpsHoisted = 0;
+  if (guardedLoop)
+    for (Operation *op : opsToMove) {
+      loop.moveOutOfLoop(op);
+      ++numOpsHoisted;
+    }
```

Comment:
as we now early exited.
```suggestion
  createLoopGuard(loop);
  size_t numOpsHoisted = 0;
  for (Operation *op : opsToMove) {
    loop.moveOutOfLoop(op);
    ++numOpsHoisted;
  }
```

Reply:
To address this comment I created a set of classes to create a loop guard for each  kind of loops we handle. 
These classes share a common base virtual class `LoopGuardBuilder` which can be used to guard an `LoopLikeOpInterface loop`:
```
LoopGuardBuilder::create(loop)->guardLoop();
```